src/recon_buildblock/LogcoshPrior.cxx

```diff
@@ -0,0 +1,666 @@
+          }
+        }
+      }
+
+      info(boost::format("parabolic_surrogate_curvature max %1%, min %2%\n") % parabolic_surrogate_curvature.find_max() % parabolic_surrogate_curvature.find_min());
+      /*{
+       static int count = 0;
+       ++count;
+       char filename[20];
+       sprintf(filename, "normalised_gradient%d.v",count);
+       write_basic_interfile(filename, parabolic_surrogate_curvature);
+       }*/
+    }
+
+    template <typename elemT>
+    Succeeded
+    LogcoshPrior<elemT>::
+    add_multiplication_with_approximate_Hessian(DiscretisedDensity<3,elemT>& output,
+                                                const DiscretisedDensity<3,elemT>& input) const
+    {
+      // TODO this function overlaps enormously with parabolic_surrogate_curvature
+      // the only difference is that parabolic_surrogate_curvature uses input==1
+
+      assert( output.has_same_characteristics(input));
+      if (this->penalisation_factor==0)
+      {
+        return Succeeded::yes;
+      }
+
+      DiscretisedDensityOnCartesianGrid<3,elemT>& output_cast =
+              dynamic_cast<DiscretisedDensityOnCartesianGrid<3,elemT> &>(output);
+
+      if (weights.get_length() ==0)
+      {
+        compute_weights(weights, output_cast.get_grid_spacing(), this->only_2D);
+      }
+
+      const bool do_kappa = !is_null_ptr(kappa_ptr);
+
+      if (do_kappa && !kappa_ptr->has_same_characteristics(input))
+        error("LogcoshPrior: kappa image has not the same index range as the reconstructed image\n");
+
+      const int min_z = output.get_min_index();
+      const int max_z = output.get_max_index();
+      for (int z=min_z; z<=max_z; z++)
+      {
+        const int min_dz = max(weights.get_min_index(), min_z-z);
+        const int max_dz = min(weights.get_max_index(), max_z-z);
+
+        const int min_y = output[z].get_min_index();
+        const int max_y = output[z].get_max_index();
+
+        for (int y=min_y;y<= max_y;y++)
+        {
+          const int min_dy = max(weights[0].get_min_index(), min_y-y);
+          const int max_dy = min(weights[0].get_max_index(), max_y-y);
+
+          const int min_x = output[z][y].get_min_index();
+          const int max_x = output[z][y].get_max_index();
+
+          for (int x=min_x;x<= max_x;x++)
+          {
+            const int min_dx = max(weights[0][0].get_min_index(), min_x-x);
+            const int max_dx = min(weights[0][0].get_max_index(), max_x-x);
+
+            elemT result = 0;
+            for (int dz=min_dz;dz<=max_dz;++dz)
+              for (int dy=min_dy;dy<=max_dy;++dy)
+                for (int dx=min_dx;dx<=max_dx;++dx)
+                {
+                  elemT current =
+                          weights[dz][dy][dx] * input[z+dz][y+dy][x+dx];
+
+                  if (do_kappa)
+                    current *=
+                            (*kappa_ptr)[z][y][x] * (*kappa_ptr)[z+dz][y+dy][x+dx];
+                  result += current;
```

Comment:
QP code?

Reply:
Would that be in the method add_multiplication_with_approximate_hessian?

In this case, if I’m not mistaken, the needed functionality should be to return
\sum_{i a neighbour of j} w_{ij}  gamma(x_i-x_j) * f_i
(to clarify, the sum goes over “i”, which are neighbours of the voxel “j”)

where x is the current image estimate and f the input (NB: in practice for SEM I will  be using f=x), and gamma(x) = rho’(x)/x, for rho being the potential function.
For the quadratic prior rho(x)=x^2/2 so gamma(x) = 1, and thus that should refer to the quadratic prior

So, in case of logcosh, ie rho(x) = delta^2 log cosh( abs(x) / delta), where delta is a predefined parameter (it is not that beta parameter that multiplies the prior), we have
gamma(x)  = delta * tanh(abs(x) / delta) / abs(x)
Defining a temp variable (as the shorthand here) I guess it should  read

temp = abs(current_image_estimate[z][y][x] - current_image_estimate[z+dz][y+dy][x+dx])/delta;
weights[dz][dy][dx] * tanh(temp)/temp * input[z+dz][y+dy][x+dx];

NB 1: there might be numerical issues if abs(x)=0, ie if the two neighbouring entries have the same value. The correct value in that case is 1, since lim_{x->0} tanh(x)/x =1, whereas if you instead use tanh(x)/(x+eps),  the value will be wrong since tanh(0) =0.
NB 2: in s(VR)EM I will also be computing 
f_j * \sum_{i a neighbour of j} w_{ij}  gamma(x_i-x_j)
where, again, f = x. As long as the method above works, this shouldn’t be a problem since I can just compute
\sum_{i a neighbour of j} w_{ij}  gamma(x_i-x_j) * 1 ——> ie by inputting ones
and then multiply the output (entry wise) with f. It also shouldn’t be that much slower I guess

Reply:
Apologies for the long comment. I can also write it in a more readable format with latex and then attach that here

Reply:
thanks. There's 2 places (easier to see on the `Files` tab) where I believe there's a copy-paste of the Quadratic prior, in the surrogate (`rho'(x)/x`), and in the Hessian (`rho''(x)`).

Would be useful to write this in function of `static inline` functions for both. Will make the clean-up easier.

NB1: yes, correct. What about using the [Taylor series for `tang`](https://www.planetmath.org/taylorseriesofhyperbolicfunctions) and say
```c++
float  surrogate(float x)
{
  if (abs(x)<eps)
  { 
   //  use Taylor: tanh(x) = x - x^3/3
   return 1- square(x)/2;
  }
  else
  {
   return tanh(x)/x;
  }
}
```
(need to put the scale factor in)

Reply:
I just noticed (ie found! the log cosh files) so NB 1 is mostly addressed in the parabolic_surrogate_curvature (where you add 0.01 to both the numerator and the denominator, giving tanh(abs(x)+0.01)/(abs(x)+0.01). this of course introduces a small error at x=0 (ie adding an epsilon introduces an error of size roughly epsilon^2/3), but that should be ok

Reply:
Apologies, I replied before seeing your comment! Yes, that should work, though I guess with square(x)/3 instead of square(x)/2?

Reply:
oops. correct

Reply:
I believe this is resolved now, can someone please check